Implement task dumps for current-thread runtime.

jswrenn opened the pull request:
## Motivation
In Java, *thread dumps* are an invaluable tool for debugging a stuck application in production. When a user requests a threaddump from a JVM, they receive text including the following information about each thread managed by the JVM:
- The thread name.
- The thread ID.
- The thread's priority.
- The thread's execution status: e.g., `RUNNABLE`, `WAITING`, or `BLOCKED`.
  This information is useful for debugging deadlocks.
- The thread's call stack.

Thread dumps provide a starting point for debugging deadlocks. A programmer can thread dump a deadlocked program, identify the `BLOCKED` threads, and inspect their call stacks to identify the resources they are blocked on.

This PR is a first step towards bringing an analogous functionality to tokio: "task dumps". A task dump is a comprehensive snapshot of the state of a Tokio application. This proposal will allow tokio services to briefly pause their work, reflect upon their internal runtime state, and comprehensively report that state to their operator.

## Solution
This PR implements a task dump mechanism for the current-thread runtime, that provides execution traces of each task. Execution traces are produced by instrumenting tokio's leaves to conditionally capture backtraces, which are then coalesced into a tree for each task.

For example, executing this program:
```rust
use std::hint::black_box;

#[inline(never)]
async fn a() {
    black_box(b()).await
}

#[inline(never)]
async fn b() {
    black_box(c()).await
}

#[inline(never)]
async fn c() {
    black_box(tokio::task::yield_now()).await
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    tokio::spawn(a());
    tokio::spawn(b());
    tokio::spawn(c());

    let handle = tokio::runtime::Handle::current();
    let dump = handle.dump();

    for (i, task) in dump.tasks().iter().enumerate() {
        let trace = task.trace();
        println!("task {i} trace:");
        println!("{trace}");
    }
}
```
Will produce an output like:
```
task 0 trace:
╼ dump::a::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:7:19
  └╼ dump::b::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:12:19
     └╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
        └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
           └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
task 1 trace:
╼ dump::b::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:12:19
  └╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
     └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
        └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
task 2 trace:
╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
  └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
     └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
```

## Limitations
- Dumps of multi-thread runtimes are `unimplemented!()`
- Dumps of current-thread runtimes must be captured from within an active runtime context.
- Dumps do not yet provide important metadata, like spawn location or task name.
- Dumps do not yet include tasks in [`LocalSet`](https://docs.rs/tokio/1.27.0/tokio/task/struct.LocalSet.html)s or [`FuturesUnordered`](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html).
- Execution traces cannot yet be decorated with extra information, like function arguments.

These limitations will be addressed in future PRs.

**NOTE:** The initial state of this PR *only* instruments `yield_now`. I will similarly instrument other leaf futures after this PR passes initial review.

## Related Reading
- https://github.com/tokio-rs/tokio/issues/5457
  This PR addresses this feature proposal.
- [***📄 Tack Trace Mechanisms***](https://hackmd.io/@jswrenn/SkldX98Ci)
  This document contrasts several potential approaches to tracing future state. This PR implements "📦 Backtrace the Leaves".

jswrenn commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,215 @@
+                                frames.push(frame.to_owned().into());
+                            }
+
+                            if ptr::eq(frame.symbol_address(), Self::leaf as *const _) {
```
The tests are failing on MacOS because the unwinder never produces a symbol address exactly equalling `Self::leaf` — it's off by twelve bytes. I don't yet know why this occurs, or if the 12 byte offset is reasonably consistent or an artifact of codegen.

jswrenn replied:
I've opened an issue on backtrace-rs seeking further clarification: https://github.com/rust-lang/backtrace-rs/issues/520.

jswrenn replied:
The issue occurs because:
> On macOS the function to get the address of the enclosing function of an ip address (`_Unwind_FindEnclosingFunction`) is unreliable due to compact unwind info collapsing multiple functions with identical unwind info together

...and is thus not used. Consequently, unwinding on macOS with `backtrace-rs` currently always emits the same value for symbol address and instruction pointer.

There are a few ways we might work around this (or, even, fix the issue in `backtrace-rs`), but in the short term I think we should probably just drop macOS support for taskdumping from this initial PR.

Darksonn commented on tokio/src/runtime/scheduler/current_thread.rs:
```diff
@@ -377,6 +377,46 @@ impl Handle {
+            // notify each task
+            let mut tasks = vec![];
+            self.shared.owned.for_each(|task| {
+                // set the notified bit
+                let _ = task.as_raw().state().transition_to_notified_for_tracing();
+                // store the raw tasks into a vec
+                tasks.push(task.as_raw());
```
I would probably encapsulate this, because I don't want ref-count logic to happen outside of `src/runtime/task`.

jswrenn replied:
Resolved by https://github.com/tokio-rs/tokio/pull/5608/commits/88ef47e962cb22cc967ce502f9371fc731dc745e?

Darksonn commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,215 @@
+    collector: Cell<Option<Trace>>,
+}
+
+/// A [`Frame`] in an intrusive, doubly-linked tree of [`Frame`]s.
+struct Frame {
+    /// The location associated with this frame.
+    inner_addr: *const c_void,
+
+    /// The parent frame, if any.
+    parent: Option<NonNull<Frame>>,
+}
```
What is the purpose of this linked tree? Is there any reason we can't just use a `Vec<*const c_void>` in the context to store this information?

Darksonn commented on tokio/src/runtime/scheduler/current_thread.rs:
```diff
@@ -377,6 +377,46 @@ impl Handle {
+        // todo: how to make this work outside of a runtime context?
+        CURRENT.with(|maybe_context| {
+            // drain the local queue
+            let Some(context) = maybe_context else { return };
+            let mut maybe_core = context.core.borrow_mut();
+            let Some(core) = maybe_core.as_mut() else { return };
```
Let-else syntax is not stabilized in our MSRV.

Darksonn commented on tokio/src/runtime/dump.rs:
```diff
@@ -0,0 +1,69 @@
+use std::fmt;
+
+/// A snapshot of a runtime's state.
+#[derive(Debug)]
+
+pub struct Dump {
+    tasks: Tasks,
+}
```
There appears to be a formatting issue here, but for some reason it is not caught by rustfmt?

Darksonn commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,215 @@
+use backtrace::BacktraceFrame;
+use std::future::Future;
+use std::{
+    cell::Cell,
+    ffi::c_void,
+    fmt,
+    pin::Pin,
+    ptr::{self, NonNull},
+    task::{self, Poll},
+};
```
Tokio uses the following import style:
```suggestion
use std::cell::Cell;
use std::ffi::c_void;
use std::fmt;
use std::pin::Pin;
use std::ptr::{self, NonNull};
use std::task::{self, Poll};
```

carllerche commented on tokio/src/util/linked_list.rs:
```diff
@@ -294,6 +294,32 @@ cfg_io_readiness! {
 
+cfg_taskdump! {
+    impl<T: Link> LinkedList<T, T::Target> {
+        pub(crate) fn for_each<F>(&mut self, mut f: F)
```
It isn't obvious why this method uses `pop_`. It looks like iteration, why is it mutating the list to iterate??

jswrenn replied:
I took a stab at writing an iterator over `LinkedList` and it segfaulted — so I shifted gears to see if I could implement `for_each` completely safely instead using the existing API.

Darksonn replied:
Please try this:
```rs
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        f(&mut *curr.as_ptr());
        next = T::pointers(curr).as_ref().get_next();
    }
}
```

jswrenn replied:
@Darksonn `&*curr.as_ptr()` gives a `&T::Target`, but `f` consumes a `&T::Handle`.

Darksonn replied:
Okay, try this:
```Rust
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        let handle = ManuallyDrop::new(T::from_raw(curr));
        f(&handle);
        next = T::pointers(curr).as_ref().get_next();
    }
}
```

carllerche commented on tokio/src/runtime/task/trace/tree.rs:
```diff
@@ -0,0 +1,126 @@
+use std::collections::{hash_map::DefaultHasher, HashMap as Map, HashSet as Set};
```
We tend only to rename imports if there is a name conflict. I was reading the code and did not recognize the `Map` and `Set` types. `HashSet` and `HashMap` let me guess the type without looking at the import list.

carllerche commented on tokio/src/runtime/scheduler/current_thread.rs:
```diff
@@ -377,6 +377,51 @@ impl Handle {
+
+        let mut traces = vec![];
+
+        // todo: how to make this work outside of a runtime context?
```
You will most likely have to send a signal somehow.

carllerche commented on tokio/src/runtime/task/state.rs:
```diff
@@ -4,7 +4,7 @@ use std::fmt;
 use std::usize;
 
-pub(super) struct State {
+pub(crate) struct State {
```
Why is this made `pub(crate)`? I don't see it used anywhere.

jswrenn replied:
Whoops, that fell through the cracks of an earlier revision. I'll correct it.

carllerche commented on tokio/src/runtime/task/raw.rs:
```diff
@@ -6,7 +6,7 @@ use std::ptr::NonNull;
 
 /// Raw task handle
-pub(super) struct RawTask {
+pub(crate) struct RawTask {
```
Why is this made `pub(crate)`. It is used by `tokio::runtime::task::trace` which should be covered by `pub(super)`.

carllerche commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,246 @@
+    let _ = injection.drain(..);
+
+    // notify each task
+    let mut tasks = vec![];
```
As an initial implementation, this is fine. Before stabilizing, we should consider how this will work when many tasks spawned onto the runtime.

carllerche reviewed (commented):
Thanks for doing this. I left several comments that are not required to apply before merging the PR, but we should create a "stabilization" tracking issue where we can keep track of remaining open questions.

jswrenn commented:
A bunch of CI jobs are stuck on "Expected — Waiting for status to be reported". Any idea why?

Darksonn commented:
That's due to a CI change on master but not on your branch. If you merge master into your branch, then the issue should go away.

Darksonn commented on .github/workflows/ci.yml:
```diff
@@ -189,6 +189,7 @@ jobs:
         os:
           - windows-latest
           - ubuntu-latest
+            rustflags: --cfg tokio_taskdump
           - macos-latest
```
Actually, it seems like the CI isn't running because this is invalid syntax.

jswrenn commented:
I believe I've addressed all outstanding issues blocking this PR. I've also added a ["Progress" section to the feature request](https://github.com/tokio-rs/tokio/issues/5457#:~:text=rs%3A50%3A13-,Progress) to track progress on the task dump feature.

Darksonn commented on tokio/src/lib.rs:
```diff
@@ -487,6 +487,23 @@ compile_error!("Tokio's build script has incorrectly detected wasm.");
+
+#[cfg(all(
+    tokio_unstable,
+    tokio_taskdump,
+    feature = "rt",
```
I would trigger this compile error even without `rt`. You aren't going to be enabling `--cfg tokio_taskdump` without knowing about it. Possibly also compile-fail on `--cfg tokio_taskdump` without rt.

jswrenn replied:
> Possibly also compile-fail on `--cfg tokio_taskdump` without rt.

Hm, unfortunately, this causes both a [CI failure](https://github.com/tokio-rs/tokio/actions/runs/4745864579/jobs/8428808277?pr=5608#step:7:81) an a usability issue: You might just keep `--cfg tokio_unstable --cfg tokio_taskdump` in your environment (.cargo/config makes this easy to do), but *not* always want to actually use the taskdumping feature. In such scenarios, your build will fail because of the explicit `compile_error!`.

Darksonn commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,246 @@
+    injection: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
+) -> Vec<Trace> {
+    // drain the local and injection queues
+    let _ = local.drain(..);
+    let _ = injection.drain(..);
```
```suggestion
    local.clear();
    injection.clear();
```

Darksonn commented on tokio/src/runtime/handle.rs:
```diff
@@ -252,6 +252,14 @@ impl Handle {
     /// [`tokio::time`]: crate::time
     #[track_caller]
     pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+        #[cfg(all(
```
Similarly, many of these cfgs could be reduced to a single `#[cfg(tokio_taskdump)]` since `lib.rs` contains a `compile_error!` if you have `#[cfg(tokio_taskdump)]` without the other options also set.

jswrenn replied:
I'm concerned that reducing these to just `#[cfg(tokio_taskdump)]` wouldn't prevent errors in these locations under odd configurations, and thus would drown out the lone explicit `compile_error` in `lib.rs`. I think Carl's suggestion is probably the way to go.

Darksonn commented on .github/workflows/ci.yml:
```diff
@@ -505,7 +508,7 @@ jobs:
         run: cargo test --no-run --lib --features full
         working-directory: tokio
         env:
-          RUSTFLAGS: --cfg loom --cfg tokio_unstable -Dwarnings
+          RUSTFLAGS: --cfg loom --cfg tokio_unstable --cfg tokio_taskdump -Dwarnings
```
You don't actually set this when you run loom tests, so I wouldn't set it when compiling them either.

Darksonn commented on .github/workflows/ci.yml:
Is `--cfg tokio_unstable` without `--cfg tokio_taskdump` still tested?

jswrenn replied:
Resolved by https://github.com/tokio-rs/tokio/pull/5608/commits/b0b11602e2864b3fd7a8f73e1c0549785aceb228

nicolasmendoza commented on tokio/src/runtime/task/trace/mod.rs:
```diff
@@ -0,0 +1,246 @@
+}
+
+/// Trace and poll all tasks of the current_thread runtime.
+pub(in crate::runtime) fn trace_current_thread(
```
Could we make the **trace_current_thread** function more readable and easier to understand? I think we should encapsulate some of the logic into separate functions to improve the function's clarity and maintainability.
Something like this (no tested)

```rust
/// Notify all tasks and return a vector of their raw pointers.
fn notify_all_tasks(owned: &OwnedTasks<Arc<current_thread::Handle>>) -> Vec<NonNull<task::RawTask>> {
    let mut tasks = vec![];
    owned.for_each(|task| {
        task.as_raw().state().transition_to_notified_for_tracing();
        tasks.push(task.as_raw());
    });
    tasks
}

/// Collect the traces for all tasks and return a vector of `Trace` objects.
fn collect_task_traces(tasks: Vec<NonNull<task::RawTask>>) -> Vec<Trace> {
    tasks
        .into_iter()
        .map(|task| {
            let ((), trace) = Trace::capture(|| task.poll());
            trace
        })
        .collect()
}

/// Trace and poll all tasks of the current_thread runtime.
pub(in crate::runtime) fn trace_current_thread(
    owned: &OwnedTasks<Arc<current_thread::Handle>>,
    local: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
    injection: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
) -> Vec<Trace> {
    local.clear();
    injection.clear();
    let tasks = notify_all_tasks(owned);
    collect_task_traces(tasks)
}
```

nicolasmendoza commented on tokio/src/runtime/dump.rs:
```diff
@@ -0,0 +1,66 @@
+
+impl Tasks {
+    /// Iterate over tasks.
+    pub fn iter(&self) -> impl Iterator<Item = &Task> {
```
One possible improvement here is to use impl Trait to simplify the definition of methods that return iterators. For example:

```suggestion
    pub fn iter(&self) -> impl Iterator<Item = &Task> + '_ {
```

**'_** means that the lifetime of the returned iterator is bound to the lifetime of the Tasks object

nicolasmendoza reviewed (commented):
GTM (Looks good to me), just some comments and minor suggestions.

Darksonn commented on .github/workflows/ci.yml:
```diff
@@ -377,7 +404,7 @@ jobs:
       - name: check --feature-powerset --unstable
         run: cargo hack check --all --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
         env:
-          RUSTFLAGS: --cfg tokio_unstable -Dwarnings
+          RUSTFLAGS: --cfg tokio_unstable --cfg tokio_taskdump -Dwarnings
```
Since this is the test designed to test various combinations of flags, I suggest we also try with just `--cfg tokio_unstable`.
```yml
      # Try with unstable feature flags
      - name: check --feature-powerset --unstable
        run: cargo hack check --all --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
        env:
          RUSTFLAGS: --cfg tokio_unstable -Dwarnings
      # Try with unstable and taskdump feature flags
      - name: check --feature-powerset --unstable --taskdump
        run: cargo hack check --all --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
        env:
          RUSTFLAGS: --cfg tokio_unstable --cfg tokio_taskdump -Dwarnings
```

Darksonn commented on tokio/src/task/yield_now.rs:
```diff
@@ -46,6 +46,15 @@ pub async fn yield_now() {
+                target_os = "linux",
+                any(target_arch = "aarch64", target_arch = "x86", target_arch = "x86_64")
+            ))]
+            crate::runtime::task::trace::Trace::leaf();
```
This PR currently only marks `yield_now` as a leaf. That's fine, but I would like more futures to be included in the first release with taskdump support.

Since we just made a Tokio release, I think we can merge this now, since we will have plenty of time to add additional functions as leaves.
